mem: Change long in src/mem/physical.cc to int64_t

erin-le opened the pull request:
This changes `long`s in src/mem/physical.cc, which are 32 bits or more, to `uint64_t`s, which are exactly 64 bits.

Change-Id: I64e089a2ac087bcf58b9c3c918c59dc5ff75d010

erin-le committed:
mem: Change long in src/mem/physical.cc to int64_t

This changes `long`s in src/mem/physical.cc, which are 32 bits
or more, to `uint64_t`s, which are exactly 64 bits.

Change-Id: I64e089a2ac087bcf58b9c3c918c59dc5ff75d010

giactra commented:
I would use Addr instead, which is still 64 bit. This is what we have been commonly using to describe memory (addresses)

powerjg commented on src/mem/physical.cc:
```diff
@@ -463,21 +463,21 @@ PhysicalMemory::unserializeStore(CheckpointIn &cp)
     uint64_t curr_size = 0;
-    long* temp_page = new long[chunk_size];
-    long* pmem_current;
+    int64_t* temp_page = new int64_t[chunk_size];
```
@giactra do you think it makes sense to use Addr here?

giactra replied:
Honestly, I don't know why we are not allocating with bytes here (using uint8_t)

powerjg replied:
Yes! That makes more sense. To answer your question, it's because this code was committed in 2006, lol

powerjg commented on src/mem/physical.cc:
```diff
@@ -463,21 +463,21 @@ PhysicalMemory::unserializeStore(CheckpointIn &cp)
+        assert(bytes_read % sizeof(int64_t) == 0);
 
-        for (uint32_t x = 0; x < bytes_read / sizeof(long); x++) {
+        for (uint32_t x = 0; x < bytes_read / sizeof(int64_t); x++) {
```
Same here.

I would say uint64_t makes the list sense, but I don't think it matters.

erin-le replied:
What do you mean by "list sense"?

powerjg replied:
See Giacomo's comment above. It was a typo since I was on my phone.

powerjg commented on src/mem/physical.cc:
```diff
@@ -367,7 +367,7 @@ PhysicalMemory::serializeStore(CheckpointOut &cp, unsigned int store_id,
     std::string filename =
         name() + ".store" + std::to_string(store_id) + ".pmem";
-    long range_size = range.size();
+    int64_t range_size = range.size();
```
Addr makes sense for range size

powerjg reviewed (commented):
The best option is a mix of Addr and uint64_t IMO. However, I worry slightly that the previous implementation was signed

erin-le committed:
mem: Allocate memory with uint8_t, use Addr in some places

This commit changes temp_page and pmem_current in
src/mem/physical.cc to be uint8_t* instead of long* or int64_t*.
It also changes range_size so it is an Addr rather than an int64_t
or long.

Change-Id: I6d97aeba0804c9ece64c76e7776de786b02e1241

giactra reviewed (changes requested):
Hi all, I gave a second look at the code, and I think I figured out the motivation behind using long here when allocating an intermediate page/buffer. The reason is *in theory* efficiency; the code down below implements a custom "optimized" memcopy. 

        for (uint32_t x = 0; x < bytes_read / sizeof(long); x++) {
            // Only copy bytes that are non-zero, so we don't give
            // the VM system hell
            if (*(temp_page + x) != 0) {
                pmem_current = (long*)(pmem + curr_size + x * sizeof(long));
                *pmem_current = *(temp_page + x);
            }

The optimization comes from the elimination of zero copies (see comment) and from the use of word sized stores to copy data (wider support than bytes sized stores) : data is copied word by word and a word is skipped if it is zero.

It is my personal opinion we should amend this code to make it more maintainable. 
We might think about replacing the for loop with a memcpy (these routines are anyway heavily optimized according to the running host), but I believe we could get rid of the intermediate copy straight away and gzread directly on pmem, updating the pointer according to the bytes_read. Please let me know your thoughts

powerjg commented:
That's great with me! We should double check the performance of restoring a checkpoint isn't significantly different though.

erin-le committed:
mem: Use long when allocating intermediate page

This commit implements giactra's code for allocating intermediate
pages and buffers. The variables temp_page and pmem_current are
changed back to the long* datatype.

Change-Id: I9323fdb82cf720b387be967ebb96de1684028f3e

erin-le committed:
mem: remove intermediate copy in src/mem/physical.cc

This commit removes the intermediate page copy in src/mem/physical.cc
and directly performs gzread with the pointer to the buffer.

Change-Id: I860e6800a7a5b7eb2768d062c9e410fafe3623a6

giactra commented on src/mem/physical.cc:
```diff
@@ -463,29 +463,17 @@ PhysicalMemory::unserializeStore(CheckpointIn &cp)
-                *pmem_current = *(temp_page + x);
-            }
-        }
+        assert(bytes_read % sizeof(int64_t) == 0);
```
I don't think this is needed anymore

giactra reviewed (changes requested):
Thanks for your changes, this looks much cleaner! I only have 2 minor comments. One is below; the other one is that we can remove chunk_size as it is not needed anymore.

As a final request, I just wanted to make sure you ran some checkpoint tests to verify cp restore still works as expected (I don't see an issue in the code per se but it's better to test this before it gets merged).

Many thanks!

erin-le commented on src/mem/physical.cc:
```diff
@@ -463,29 +463,17 @@ PhysicalMemory::unserializeStore(CheckpointIn &cp)
     uint32_t bytes_read;
     while (curr_size < range.size()) {
-        bytes_read = gzread(compressed_mem, temp_page, chunk_size);
+        bytes_read = gzread(compressed_mem, pmem, chunk_size);
```
It looks like chunk_size is still used here. Since this is the only place it's used, should I put the value of chunk_size directly into gzread and remove the declaration, or would it be better to leave it as is?

erin-le committed:
mem: Remove unnecessary assert in physical.cc

This commit removes an assert in src/mem/physical.cc that became
obsolete as a result of other changes.

Change-Id: If2e3525426060bf6ed28810092d57cb66f6169a0

erin-le commented:
> As a final request, I just wanted to make sure you ran some checkpoint tests to verify cp restore still works as expected (I don't see an issue in the code per se but it's better to test this before it gets merged).

For testing, I used `riscv-hello-restore-checkpoint.py` and `simpoints-se-restore.py` from gem5_library and ran them using both gem5 stable and the build for this branch. The diff for stats.txt only showed slight differences with stats such as hostTickRate, hostMemory, hostSeconds, etc., and the diff for standard out only showed that I was running different gem5 versions. Based on this, I believe it should still work properly.
